mypyc/test-data/run-classes.test

```diff
@@ -3277,3 +3277,242 @@ def test_python_class() -> None:
 
 [file dynamic.py]
 class Dyn: pass
+
+[case testDunderNew]
+from __future__ import annotations
+from typing import Union
+
+class Add:
+    l: IntLike
+    r: IntLike
+
+    def __new__(cls, l: IntLike, r: IntLike) -> IntLike:  # type: ignore[misc]
+        print(f'running __new__ with {l} and {r}')
+
+        return (
+            l if r == 0 else
+            r if l == 0 else
+            super().__new__(cls)
+        )
+
+    def __init__(self, l: IntLike, r: IntLike):
+        self.l = l
+        self.r = r
+
+    def __repr__(self) -> str:
+        return f'({self.l} + {self.r})'
+
+    def __add__(self, other: IntLike) -> IntLike:
+        return Add(self, other)
+
+IntLike = Union[int, Add]
+
+[file driver.py]
+from native import Add
```

Comment:
This is only testing behavior in interpreted code. What about also testing that similar things work in compiled code?